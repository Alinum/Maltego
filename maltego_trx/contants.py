BOOKMARK_COLOR_NONE = "-1"
BOOKMARK_COLOR_BLUE = "0"
BOOKMARK_COLOR_GREEN = "1"
BOOKMARK_COLOR_YELLOW = "2"
BOOKMARK_COLOR_PURPLE = "3"
BOOKMARK_COLOR_RED = "4"
BOOKMARK_CLRS = {
        "none":   BOOKMARK_COLOR_NONE,
        "blue":   BOOKMARK_COLOR_BLUE,
        "green":  BOOKMARK_COLOR_GREEN,
        "yellow": BOOKMARK_COLOR_YELLOW,
        "purple": BOOKMARK_COLOR_PURPLE,
        "red":    BOOKMARK_COLOR_RED,
}
LINK_STYLE_NORMAL = "0"
LINK_STYLE_DASHED = "1"
LINK_STYLE_DOTTED = "2"
LINK_STYLE_DASHDOT = "3"
LINK_STYLES = {
        "normal":  LINK_STYLE_NORMAL,
        "dashed":  LINK_STYLE_DASHED,
        "dotted":  LINK_STYLE_DOTTED,
        "dashdot": LINK_STYLE_DASHDOT,
}
UIM_FATAL = 'FatalError'
UIM_PARTIAL = 'PartialError'
UIM_INFORM = 'Inform'
UIM_DEBUG = 'Debug'
UIM_TYPES = {
        "fatal":   UIM_FATAL,
        "partial": UIM_PARTIAL,
        "inform":  UIM_INFORM,
        "debug":   UIM_DEBUG,
}
