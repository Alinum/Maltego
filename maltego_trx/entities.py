Alias = "maltego.Alias"
ASNumber = "maltego.AS"
BuiltwithTechnology = "maltego.BuiltwithTechnology"
BuiltWithRelationship = "maltego.BuiltWithRelationship"
CircularArea = "maltego.CircularArea"
Company = "maltego.Company"
Device = "maltego.Device"
DNS = "maltego.DNSName"
Document = "maltego.Document"
Domain = "maltego.Domain"
Email = "maltego.EmailAddress"
FlickrAffiliation = "maltego.AffiliationFlickr"
GPS = "maltego.GPS"
Hash = "maltego.Hash"
Hashtag = "maltego.Hashtag"
Image = "maltego.Image"
IPAddress = "maltego.IPv4Address"
Location = "maltego.Location"
MX = "maltego.MXRecord"
MyspaceAffiliation = "maltego.AffiliationMyspace"
Namechk = "maltego.Namechk"
Netblock = "maltego.Netblock"
NS = "maltego.NSRecord"
Organization = "maltego.Organization"
Person = "maltego.Person"
PhoneNumber = "maltego.PhoneNumber"
Phrase = "maltego.Phrase"
Port = "maltego.Port"
Sentiment = "maltego.Sentiment"
StockSymbol = "maltego.StockSymbol"
Tweet = "maltego.Twit"
TwitterAffiliation = "maltego.AffiliationTwitter"
TwitterList = "maltego.TwitterUserList"
UniqueID = "maltego.UniqueIdentifier"
URL = "maltego.URL"
Website = "maltego.Website"
WebTitle = "maltego.WebTitle"