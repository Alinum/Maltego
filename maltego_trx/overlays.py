from enum import Enum


class Position(Enum):
    NORTH = "N"
    NORTH_EAST = "NE"
    NORTH_WEST = "NW"
    EAST = "E"
    CENTER = "C"
    WEST = "W"
    SOUTH = "S"
    SOUTH_EAST = "SE"
    SOUTH_WEST = "SW"


class OverlayType(Enum):
    IMAGE = "image"
    COLOUR = "colour"
    TEXT = "text"
