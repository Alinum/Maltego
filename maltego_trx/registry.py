import importlib
import logging
import os
import pkgutil
import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import chain
from typing import List, Literal, Optional, Dict, Generator, Sized

TRANSFORMS_CSV_HEADER = "Owner,Author,Disclaimer,Description,Version,Name,UIName,URL,entityName,oAuthSettingId," \
                        "transformSettingIDs,seedIDs"
SETTINGS_CSV_HEADER = "Name,Type,Display,DefaultValue,Optional,Popup"


@dataclass()
class TransformMeta:
    class_name: str
    display_name: str
    input_entity: str
    description: str


@dataclass()
class TransformSetting:
    name: str
    setting_type: Literal['string', 'boolean', 'date', 'datetime', 'daterange', 'url', 'double', 'int']
    display_name: str
    default_value: Optional[str] = None
    optional: bool = False
    popup: bool = False
    global_setting: bool = False

    @property
    def fullname(self) -> str:
        if self.global_setting:
            return "global#" + self.name
        return self.name


def chunk_list(data: Sized, max_chunk_size: int) -> Generator[list, None, None]:
    # rather use smaller, even lists than maxed out lists and a small tail
    number_of_chunks = len(data) // max_chunk_size + 1

    # if it's a perfect split, do it
    if len(data) % max_chunk_size == 0:
        number_of_chunks -= 1

    chunk_size = len(data) // number_of_chunks + 1

    for i in range(0, len(data), chunk_size):
        yield data[i:i + chunk_size]


@dataclass()
class TransformRegistry:
    owner: str
    author: str

    disclaimer: str
    host_url: str
    seed_ids: List[str]

    version: str = '0.1'
    prefix: str = ""
    suffix: str = ""

    _settings: Dict[str, TransformSetting] = field(default_factory=lambda: defaultdict(list))
    global_settings: List[TransformSetting] = field(default_factory=list)
    settings_prefix: str = field(default="")

    oauth_settings_id: Optional[str] = field(default="")

    metas: List[TransformMeta] = field(init=False, default_factory=list)
    transform_settings: Dict = field(default_factory=dict)

    transform_classes: dict[str, object] = field(default_factory=dict)

    @staticmethod
    def scan_for_transforms(*modules):
        """This method just imports modules recursively to run the decorators 'register' and 'uses_settings'"""
        for module in modules:
            prefix = module.__name__ + "."  # transform.
            for importer, modname, ispkg in pkgutil.walk_packages(module.__path__, prefix):
                if not ispkg:
                    importlib.import_module(modname)

    @property
    def settings(self) -> List[TransformSetting]:
        return list(self._settings.values())

    @settings.setter
    def settings(self, settings: List[TransformSetting]):
        self._settings = {setting.name: setting for setting in settings}

    @staticmethod
    def pascal_case_to_title(name: str):
        # https://stackoverflow.com/a/1176023
        name = re.sub('(.)([A-Z][a-z]+)', r'\1 \2', name)
        return re.sub('([a-z0-9])([A-Z])', r'\1 \2', name)

    def register(self, display_name: str, input_entity: str, description: str):
        def decorated(transform_class: object):
            self.transform_classes[transform_class.__name__.lower()] = transform_class

            display = display_name or self.pascal_case_to_title(transform_class.__name__)
            self.metas.append(TransformMeta(transform_class.__name__, display, input_entity, description))

            return transform_class

        return decorated

    def uses_settings(self, *settings: str):
        def decorated(transform_class: object):
            self.transform_settings[transform_class.__name__] = list(settings)

            return transform_class

        return decorated

    def write_transforms_config(self, config_path: str = "./transforms.csv", max_transforms_per_file: int = 100):
        lines = []

        global_settings_fullnames = [gs.fullname for gs in self.global_settings]

        for transform_meta in self.metas:
            meta_settings = list()
            for setting in self.transform_settings.get(transform_meta.class_name, []):
                if transform_setting := self._settings.get(setting):
                    meta_settings.append(transform_setting.fullname)
                else:
                    logging.warning(
                            f"Setting {setting} is not registered. Please add to {self.__class__.__name__}.settings")

            # combine global and transform-scoped settings
            datum = [
                    self.owner,
                    self.author,
                    self.disclaimer,
                    transform_meta.description,
                    self.version,
                    transform_meta.class_name,
                    self.prefix + transform_meta.display_name + self.suffix,
                    os.path.join(self.host_url, "run", transform_meta.class_name),
                    transform_meta.input_entity,
                    self.oauth_settings_id,
                    ";".join(chain(meta_settings, global_settings_fullnames)),
                    ";".join(self.seed_ids)
            ]

            lines.append(",".join(datum))

        if len(lines) <= max_transforms_per_file:
            with open(config_path, "w+") as transforms_config:
                transforms_config.write(TRANSFORMS_CSV_HEADER + "\n")
                transforms_config.writelines(map(lambda x: x + "\n", lines))

            return

        # split file to speed-up import
        chunks = list(chunk_list(lines, max_transforms_per_file))
        for idx, chunk in enumerate(chunks):
            path, extension = config_path.rsplit(".", 1)
            chunked_config_path = f"{path}_{idx}-{len(chunks)}.{extension}"

            with open(chunked_config_path, "w+") as transforms_config:
                transforms_config.write(TRANSFORMS_CSV_HEADER + "\n")
                transforms_config.writelines(map(lambda x: x + "\n", chunk))

    def write_settings_config(self, config_path: str = "./settings.csv"):
        settings = []

        for setting in chain(self.global_settings, self.settings):
            settings.append(",".join([
                    setting.name,
                    setting.setting_type,
                    setting.display_name,
                    setting.default_value or "",
                    "true" if setting.optional else "false",
                    "true" if setting.popup else "false"
            ]))

        with open(config_path, "w+") as settings_config:
            settings_config.write(SETTINGS_CSV_HEADER + "\n")
            settings_config.writelines(map(lambda x: x + "\n", settings))
