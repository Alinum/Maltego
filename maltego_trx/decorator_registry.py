import os
import sys
import typing
import zipfile
from dataclasses import dataclass, field
from itertools import chain
from typing import List, Optional, Dict, Iterable, Type, Set
from xml.etree import ElementTree
from xml.etree.ElementTree import Element

from maltego_trx.protocol.entities import MaltegoEntityType, MaltegoEntityTypeMeta, generate_meta, EntityFieldType, \
    generate_field
from maltego_trx.protocol.mtz import create_entity_type_xml
from maltego_trx.utils import filter_unique, pascal_case_to_title, escape_csv_fields, export_as_csv, serialize_bool, \
    name_to_path

TRANSFORMS_CSV_HEADER = "Owner,Author,Disclaimer,Description,Version," \
                        "Name,UIName,URL,entityName," \
                        "oAuthSettingId,transformSettingIDs,seedIDs"
SETTINGS_CSV_HEADER = "Name,Type,Display,DefaultValue,Optional,Popup"


@dataclass()
class TransformMeta:
    class_name: str
    display_name: str
    input_entity: str
    description: str
    output_entities: List[str]
    disclaimer: str


@dataclass()
class TransformSetting:
    name: str
    display_name: str
    setting_type: str  # Literal['string', 'boolean', 'date', 'datetime', 'daterange', 'url', 'double', 'int']

    default_value: Optional[str] = ""
    optional: bool = False
    popup: bool = False
    global_setting: bool = False

    @property
    def id(self) -> str:
        """this setting's full id for reference"""
        if self.global_setting:
            return "global#" + self.name
        return self.name


@dataclass(eq=False)
class TransformRegistry:
    owner: str
    author: str

    host_url: str
    seed_ids: List[str]

    version: str = '0.1'
    display_name_suffix: str = ""

    global_settings: List[TransformSetting] = field(default_factory=list)
    oauth_settings_id: Optional[str] = ""

    namespace: str = None

    transform_metas: Dict[str, TransformMeta] = field(init=False, default_factory=dict)
    transform_settings: Dict[str, List[TransformSetting]] = field(init=False, default_factory=dict)

    entity_metas: Dict[str, Type[MaltegoEntityType]] = field(init=False, default_factory=dict)
    entity_categories: Set[str] = field(init=False, default_factory=set)

    def register_transform(self, display_name: str, input_entity: str, description: str,
                           settings: List[TransformSetting] = None, output_entities: List[str] = None,
                           disclaimer: str = ""):
        """ This method can be used as a decorator on transform classes. The data will be used to fill out csv config
            files to be imported into a TDS.
        """

        def decorated(transform_callable: object):
            cleaned_transform_name = name_to_path(transform_callable.__name__)
            display = display_name or pascal_case_to_title(transform_callable.__name__)

            meta = TransformMeta(cleaned_transform_name,
                                 display, input_entity,
                                 description,
                                 output_entities or [],
                                 disclaimer)
            self.transform_metas[cleaned_transform_name] = meta

            if settings:
                self.transform_settings[cleaned_transform_name] = settings

            return transform_callable

        return decorated

    def register_entity(self,
                        namespace: str = None,
                        name: str = None,
                        display_name: str = None,
                        display_name_plural: str = None,
                        description: str = None,
                        category: str = None,
                        small_icon_resource: str = None,
                        large_icon_resource: str = None,
                        allowed_root: bool = True,
                        conversion_order: int = 2147483647,
                        visible: bool = True,
                        base_entities: list[str] = None,
                        value_field: str = None,
                        display_value_field: str = None):
        # get all arguments
        kwargs = locals()
        kwargs.pop("self")
        kwargs['namespace'] = kwargs['namespace'] or self.namespace

        def decorated(cls: Type[MaltegoEntityType]) -> MaltegoEntityType:

            default = MaltegoEntityTypeMeta(**kwargs)
            meta = getattr(cls, '__meta__', default)
            meta = generate_meta(cls.__name__, meta)
            cls.__meta__ = meta

            if meta.category:
                self.entity_categories.add(meta.category)

            field_metas = {}
            for field_name, field_type in typing.get_type_hints(cls).items():
                if field_name.startswith("__"):
                    continue

                sample = None
                field_meta = None
                if value := getattr(cls, field_name, None):
                    if isinstance(value, EntityFieldType):
                        field_meta = value
                    else:
                        sample = value

                field_meta = generate_field(field_name, sample, field_meta)
                setattr(cls, field_name, field_meta)
                field_metas[field_name] = field_meta

            if not meta.value_field:
                meta.value_field = list(field_metas.keys())[0]
            if not meta.display_value_field:
                meta.display_value_field = meta.value_field

            self.entity_metas[meta.id] = cls
            cls.__fields__ = field_metas
            return cls

        return decorated

    def write_transforms_config(self, config_path: str = "./transforms.csv", csv_line_limit: int = 100):
        """Exports the collected transform meta data as a csv-file to config_path"""
        global_settings_full_names = [gs.id for gs in self.global_settings]

        csv_lines = []
        for transform_name, transform_meta in self.transform_metas.items():
            meta_settings = [setting.id for setting in
                             self.transform_settings.get(transform_name, [])]

            transform_row = [
                self.owner,
                self.author,
                transform_meta.disclaimer,
                transform_meta.description,
                self.version,
                transform_name,
                transform_meta.display_name + self.display_name_suffix,
                os.path.join(self.host_url, "run", transform_name),
                transform_meta.input_entity,
                ";".join(self.oauth_settings_id),
                # combine global and transform scoped settings
                ";".join(chain(meta_settings, global_settings_full_names)),
                ";".join(self.seed_ids)
            ]

            escaped_fields = escape_csv_fields(*transform_row)
            csv_lines.append(",".join(escaped_fields))

        export_as_csv(TRANSFORMS_CSV_HEADER, csv_lines, config_path, csv_line_limit)

    def write_settings_config(self, config_path: str = "./settings.csv", csv_line_limit: int = 100):
        """Exports the collected settings meta data as a csv-file to config_path"""
        chained_settings = chain(self.global_settings, *list(self.transform_settings.values()))
        unique_settings: Iterable[TransformSetting] = filter_unique(lambda s: s.name, chained_settings)

        csv_lines = []
        for setting in unique_settings:
            setting_row = [
                setting.id,
                setting.setting_type,
                setting.display_name,
                setting.default_value or "",
                serialize_bool(setting.optional, 'True', 'False'),
                serialize_bool(setting.popup, 'Yes', 'No')
            ]

            escaped_fields = escape_csv_fields(*setting_row)
            csv_lines.append(",".join(escaped_fields))

        export_as_csv(SETTINGS_CSV_HEADER, csv_lines, config_path, csv_line_limit)

    def write_mtz(self, mtz_path: str):
        with zipfile.ZipFile(mtz_path, "w") as mtz:
            for name, entity_meta in self.entity_metas.items():
                entity_xml = create_entity_type_xml(entity_meta)
                if sys.version_info.minor >= 9:
                    ElementTree.indent(entity_xml)

                entity_xml_str = ElementTree.tostring(entity_xml)
                mtz.writestr(f"Entities/{name}.entity", entity_xml_str)

            for category in self.entity_categories:
                category_xml = Element("EntityCategory", attrib={"name": category})
                category_xml_str = ElementTree.tostring(category_xml)
                mtz.writestr(f"EntityCategories/{category.lower()}.category", category_xml_str)
