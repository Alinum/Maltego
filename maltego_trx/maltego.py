import uuid
from dataclasses import dataclass, field
from typing import AnyStr, TypedDict, Literal
from xml.etree import ElementTree
from xml.etree.ElementTree import Element

from .entities import translate_legacy_property_name
from .overlays import OverlayType, OverlayPosition


@dataclass()
class DisplayInfo:
    name: str
    content: str

    def build_xml(self) -> Element:
        return Element('Label', attrib={'Name': self.name, 'Content': self.content})


@dataclass()
class EntityProperty:
    name: str
    display: str
    value: str
    matching: str = field(default=None)

    def build_xml(self) -> Element:
        field_xml = Element('Field', attrib={'Name':         self.name,
                                             'DisplayName':  self.display,
                                             'MatchingRule': self.matching})

        field_xml.text = f'<![CDATA[{self.value}]]>'
        return field_xml


@dataclass()
class EntityOverlay:
    property_name: str
    position: OverlayPosition
    overlay_type: OverlayType

    def build_xml(self) -> Element:
        return Element('Overlay', attrib={'propertyName': self.property_name,
                                          'position':     self.position,
                                          'type':         self.overlay_type})


@dataclass()
class UiMessage:
    message: str
    message_type: Literal['Inform', 'Debug', 'FatalError', 'PartialError'] = field(default='Inform')

    def build_xml(self) -> Element:
        ui_message = Element('UIMessage', attrib={'MessageType': self.message_type})
        ui_message.text = self.message

        return ui_message


@dataclass()
class CleanMaltegoEntity:
    entity_type: str
    value: str

    weight: int = field(default=100)

    display_information: list[DisplayInfo] = field(default_factory=list)
    icon_url: str = field(default="")

    properties: list[EntityProperty] = field(default_factory=list)
    overlays: list[EntityOverlay] = field(default_factory=list)

    def add_properties(self, *properties: EntityProperty):
        self.properties.extend(properties)

    def add_overlays(self, *overlays: EntityOverlay):
        self.overlays.extend(overlays)

    def build_xml(self) -> Element:
        entity = Element('Entity')
        value = Element('Value')
        value.text = f'<![CDATA[{self.entity_type}]]>'

        weight = Element('Weight')
        weight.text = f'<![CDATA[{self.weight}]]>'

        display_info_xml = Element('DisplayInformation')
        for display_info in self.display_information:
            display_info_xml.append(display_info.build_xml())

        properties_xml = Element('AdditionalFields')
        for prop in self.properties:
            properties_xml.append(prop.build_xml())

        overlays_xml = Element('Overlays')
        for overlay in self.overlays:
            overlays_xml.append(overlay.build_xml())

        icon_xml = Element('IconURL')
        icon_xml.text = self.icon_url

        return entity


@dataclass()
class MaltegoResponse:
    entities: list[CleanMaltegoEntity] = field(default_factory=list)
    exceptions: list[str] = field(default_factory=list)
    ui_messages: list[UiMessage] = field(default_factory=list)

    def build_exceptions_xml(self) -> Element:
        exceptions_xml = Element('Exceptions')
        for exception in self.exceptions:
            exception_xml = Element('Exception')
            exception_xml.text = exception

            exceptions_xml.append(exception_xml)

        response_message = Element('MaltegoTransformResponseMessage')
        response_message.append(exceptions_xml)

        message = Element('MaltegoMessage')
        message.append(response_message)

        return message

    def build_xml(self) -> Element:
        entities_xml = Element('Entities')
        for entity in self.entities:
            entities_xml.append(entity.build_xml())

        ui_message_xml = Element('UIMessages')
        for ui_message in self.ui_messages:
            ui_message_xml.append(ui_message.build_xml())

        response_message = Element('MaltegoTransformResponseMessage')
        response_message.append(entities_xml)
        response_message.append(ui_message_xml)

        message = Element('MaltegoMessage')
        message.append(response_message)

        return message


class Genealogy(TypedDict):
    name: str
    old_name: str


class MaltegoRequest:
    value: str
    properties: dict[str, str]
    type: str

    weight: int
    slider: int

    genealogy: list[Genealogy]
    settings: dict[str, str]

    @staticmethod
    def _parse_entity(entity: Element, genealogy: list[Genealogy]) -> dict[str, str]:
        properties = dict()
        for prop in entity.iter("Field"):
            name = prop.attrib.get("Name")

            properties[name] = prop.text

            for entity_type in genealogy:
                v3_property_name = translate_legacy_property_name(entity_type["name"], name)
                if v3_property_name is not None:
                    properties[v3_property_name] = prop.text

        return properties

    @staticmethod
    def _parse_genealogy(genealogy_xml: Element) -> list[Genealogy]:
        genealogy = list()
        for genealogy_type in genealogy_xml.iter('Type'):
            entity_type_name = genealogy_type.attrib.get("Name")
            entity_type_old_name = genealogy_type.attrib.get("OldName", None)

            genealogy.append(Genealogy(name=entity_type_name, old_name=entity_type_old_name))

        return genealogy

    @staticmethod
    def _parse_settings(settings_xml: Element) -> dict[str, str]:
        settings = {}
        for setting in settings_xml.iter('Field'):
            name = setting.attrib.get('Name')
            settings[name] = setting.text

        return settings

    @classmethod
    def from_xml(cls, xml_string: AnyStr) -> 'MaltegoRequest':
        request: Element = ElementTree.fromstring(xml_string)

        message = MaltegoRequest()

        message.type = request.find('Type')
        message.value = request.findtext('Value')
        message.slider = int(request.findtext('Slider'))

        limit = request.find('Limits').attrib.get('SoftLimit')
        message.weight = int(limit)

        genealogy_xml = request.find('Genealogy')
        message.genealogy = cls._parse_genealogy(genealogy_xml)

        entity_xml = next(request.iter('Entity'))
        message.entity = cls._parse_entity(entity_xml, message.genealogy)

        settings_xml = request.find('TransformFields')
        message.settings = cls._parse_settings(settings_xml)

        return message

    @classmethod
    def from_local_args(cls, local_args: list) -> 'MaltegoRequest':
        message = MaltegoRequest()

        message.value = local_args[0]

        message.type = "local.Unknown"
        message.genealogy = None

        message.weight = 100
        message.slider = 100

        # don't know what this does
        if len(local_args) > 1:
            hash_rnd = uuid.uuid4().hex.upper()[0:7]
            equals_rnd = uuid.uuid4().hex.upper()[0:7]
            bslash_rnd = uuid.uuid4().hex.upper()[0:7]
            text = local_args[1] \
                .replace("\\\\", bslash_rnd) \
                .replace("\\#", hash_rnd) \
                .replace("\\=", equals_rnd)

            properties = {}
            for property_section in text.split("#"):
                name, value = property_section.split("=", 2)
                properties[name] = value \
                    .replace(hash_rnd, "#") \
                    .replace(equals_rnd, "=") \
                    .replace(bslash_rnd, "\\")

            message.properties = properties
            message.settings = {}

        return message
