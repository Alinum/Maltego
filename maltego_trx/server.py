import logging
from xml.etree.ElementTree import Element, SubElement

from flask import Flask, request

from maltego_trx.maltego import MaltegoMsg
from .registry import mapping
from .utils import serialize_xml

log = logging.getLogger("maltego.server")
logging.basicConfig(level=logging.DEBUG)

URL_TEMPLATE = '/run/<transform_name>/'
URL_TEMPLATE_NO_SLASH = '/run/<transform_name>'


def build_exception_message(msg="An exception occurred with the transform. Check the logs for more details."):
    message_xml = Element('MaltegoMessage')
    response_xml = SubElement(message_xml, 'MaltegoTransformResponseMessage')

    ui_messages_xml = SubElement(response_xml, 'UIMessages')
    ui_message_xml = SubElement(ui_messages_xml, 'UIMessage', MessageType='PartialError')
    ui_message_xml.text = msg

    return message_xml


def get_exception_message(msg="An exception occurred with the transform. Check the logs for more details."):
    return serialize_xml(build_exception_message(msg))


def print_transforms():
    print("= Transform Server URLs =")
    for path in mapping:
        print(URL_TEMPLATE.replace("<transform_name>", path) + ": " + mapping[path].__name__)
    print("\n")

    print("= Local Transform Names =")
    for path in mapping:
        print(path + ": " + mapping[path].__name__)
    print("\n")


def run_transform(transform_name, client_msg):
    transform_method = mapping[transform_name]
    try:
        if hasattr(transform_method, "run_transform"):
            return transform_method.run_transform(client_msg), 200  # Transform class
        else:
            return transform_method(client_msg), 200  # Transform method
    except Exception as e:
        log.error("An exception occurred while executing your transform code.")
        log.error(e, exc_info=True)
        return get_exception_message(), 200


app = Flask(__name__)
application = app  # application variable for usage with apache mod wsgi


def transform_runner(transform_name):
    transform_name = transform_name.lower()
    if transform_name in mapping:
        if request.method == 'POST':
            client_msg = MaltegoMsg(request.data)
            return run_transform(transform_name, client_msg)
        else:
            return "Transform found with name '%s', you will need to send a POST request to run it." % transform_name, 200
    else:
        log.info("No transform found with the name '%s'." % transform_name)
        log.info("Available transforms are:\n %s" % str(list(mapping.keys())))
        return "No transform found with the name '%s'." % transform_name, 404


# Add the route with and without the slash, since POSTs can't be redirected
app.route(URL_TEMPLATE_NO_SLASH, methods=['GET', 'POST'])(transform_runner)
app.route(URL_TEMPLATE, methods=['GET', 'POST'])(transform_runner)


@app.route('/', methods=['GET', 'POST'])
def index():
    return "You have reached a Maltego Transform Server.", 200
