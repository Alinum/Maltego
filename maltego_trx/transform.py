from abc import abstractmethod

from maltego_trx.maltego import MaltegoRequest, MaltegoResponse


class DiscoverableTransform:
    @classmethod
    @abstractmethod
    async def create_entities(cls, request: MaltegoRequest, response: MaltegoResponse):
        raise NotImplementedError("create_entities static method must be implemented in child class.")

    @classmethod
    async def run_transform(cls, request: MaltegoRequest):
        response = MaltegoResponse()
        await cls.create_entities(request, response)
        return response.build_xml()
