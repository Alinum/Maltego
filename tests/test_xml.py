from maltego_trx.maltego import MaltegoTransform
from maltego_trx.overlays import OverlayType, OverlayPosition


def test_entity_with_value(snapshot):
    response = MaltegoTransform()
    response.addEntity("maltego.Phrase", "Hello Spencer!")
    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_entity_with_properties(snapshot):
    response = MaltegoTransform()
    entity = response.addEntity("maltego.Phrase", "Hello Spencer!")
    entity.addProperty(fieldName="fieldNameTest", displayName="displayNameTest", value="valueTest",
                       matchingRule="loose")
    entity.addProperty(fieldName="fieldNameTest2", displayName="displayNameTest2", value="valueTest2",
                       matchingRule="strict")

    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_entity_with_display_information(snapshot):
    response = MaltegoTransform()
    entity = response.addEntity("maltego.Phrase", "Hello Spencer!")
    entity.addDisplayInformation(title="Display Info Title", content="<p>Display Info Content</p>")

    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_entity_with_icon(snapshot):
    response = MaltegoTransform()
    entity = response.addEntity("maltego.Phrase", "Hello Spencer!")
    entity.setIconURL(url="https://www.maltego.com/img/maltego-logo/maltego-horizontal.png")

    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_entity_with_overlays(snapshot):
    response = MaltegoTransform()
    entity = response.addEntity("maltego.Phrase", "Hello Spencer!")
    entity.addOverlay('#45e06f', OverlayPosition.NORTH_WEST, OverlayType.COLOUR)

    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_response_with_ui_message(snapshot):
    response = MaltegoTransform()
    response.addUIMessage("Test Message", messageType="inform")

    response_xml = response.returnOutput()

    assert response_xml == snapshot


def test_response_with_exception(snapshot):
    response = MaltegoTransform()
    response.addException("Test Exception")

    response_xml = response.throwExceptions()

    assert response_xml == snapshot


def test_exception_message(snapshot):
    transform_run = MaltegoTransform()
    transform_run.addUIMessage("Test Exception", "PartialError")
    response_xml = transform_run.returnOutput()

    assert response_xml == snapshot


def test_all_null_values(caplog, snapshot):
    response = MaltegoTransform()
    entity = response.addEntity(None, None)
    entity.addProperty(fieldName=None, displayName=None, value=None,
                       matchingRule=None)

    entity.addDisplayInformation(title=None, content=None)
    entity.setIconURL(url=None)

    entity.addOverlay(propertyName=None, position=OverlayPosition.NORTH_WEST, overlayType=OverlayType.COLOUR)

    entity.setLinkLabel(None)
    entity.setLinkThickness(None)
    entity.setLinkColor(None)
    entity.setLinkStyle(None)
    entity.setType(None)
    entity.setWeight(None)
    entity.addCustomLinkProperty(None, None, None)
    entity.setNote(None)
    entity.setValue(None)

    response.addUIMessage(None, messageType=None)
    response.addException(None)

    response_xml = response.returnOutput()
    assert response_xml

    captured = caplog.text
    assert captured == snapshot
