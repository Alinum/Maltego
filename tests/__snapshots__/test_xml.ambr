# name: test_all_null_values
  list([
    'Additional field is missing a title, display, matching or value: title=None, display=None, matching=None, val=None',
    'Additional field is missing a title, display, matching or value: title=link#maltego.link.label, display=Label, matching=, val=None',
    'Additional field is missing a title, display, matching or value: title=link#maltego.link.thickness, display=Thickness, matching=, val=None',
    'Additional field is missing a title, display, matching or value: title=link#maltego.link.color, display=LinkColor, matching=, val=None',
    'Additional field is missing a title, display, matching or value: title=link#maltego.link.style, display=LinkStyle, matching=, val=None',
    'Additional field is missing a title, display, matching or value: title=notes#, display=Notes, matching=, val=None',
    'Overlay is missing a property name, position or type: property_name=None, position=NW, overlay_type=colour',
    'UIMessage is missing a message type or content: message_type=None, message_content=None',
  ])
# ---
# name: test_entity_with_display_information
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities><Entity Type="maltego.Phrase"><Value>Hello Spencer!</Value><Weight>100</Weight><DisplayInformation><Label Name="Display Info Title" Type="text/html">&lt;p&gt;Display Info Content&lt;/p&gt;</Label></DisplayInformation></Entity></Entities><UIMessages></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_entity_with_icon
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities><Entity Type="maltego.Phrase"><Value>Hello Spencer!</Value><Weight>100</Weight><IconURL>https://www.maltego.com/img/maltego-logo/maltego-horizontal.png</IconURL></Entity></Entities><UIMessages></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_entity_with_overlays
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities><Entity Type="maltego.Phrase"><Value>Hello Spencer!</Value><Weight>100</Weight><Overlays><Overlay position="NW" propertyName="#45e06f" type="colour"></Overlay></Overlays></Entity></Entities><UIMessages></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_entity_with_properties
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities><Entity Type="maltego.Phrase"><Value>Hello Spencer!</Value><Weight>100</Weight><AdditionalFields><Field DisplayName="displayNameTest" MatchingRule="loose" Name="fieldNameTest">valueTest</Field><Field DisplayName="displayNameTest2" MatchingRule="strict" Name="fieldNameTest2">valueTest2</Field></AdditionalFields></Entity></Entities><UIMessages></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_entity_with_value
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities><Entity Type="maltego.Phrase"><Value>Hello Spencer!</Value><Weight>100</Weight></Entity></Entities><UIMessages></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_exception_message
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities></Entities><UIMessages><UIMessage MessageType="PartialError">Test Exception</UIMessage></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
# name: test_response_with_exception
  '<MaltegoMessage><MaltegoTransformExceptionMessage><Exceptions><Exception>Test Exception</Exception></Exceptions></MaltegoTransformExceptionMessage></MaltegoMessage>'
# ---
# name: test_response_with_ui_message
  '<MaltegoMessage><MaltegoTransformResponseMessage><Entities></Entities><UIMessages><UIMessage MessageType="inform">Test Message</UIMessage></UIMessages></MaltegoTransformResponseMessage></MaltegoMessage>'
# ---
